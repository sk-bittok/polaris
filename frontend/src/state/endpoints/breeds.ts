import type {
	RegisterBreedSchema,
	UpdateBreedSchema,
} from "@/lib/schemas/animal";
import type { Breed } from "@/models/livestock";
import type { EndpointBuilder } from "@reduxjs/toolkit/query";

export const breedEndpoints = (build: EndpointBuilder<any, any, any>) => ({
	getBreeds: build.query<Breed[], void>({
		query: () => "/breeds",
		providesTags: ["GetBreeds"],
	}),
	getBreedById: build.query<Breed, number>({
		query: (id) => ({ url: `/breeds/${id}` }),
		providesTags: (result, error, id) => [{ type: "GetBreeds", id }],
	}),
	createBreed: build.mutation<Breed, RegisterBreedSchema>({
		query: (params) => ({
			url: "/breeds",
			method: "POST",
			body: {
				specie: params.specie,
				name: params.name,
				description: params.description,
				typicalMaleWeightRange: params.maleWeightRange,
				typicalFemaleWeightRange: params.femaleWeightRange,
				typicalGestationPeriod: params.gestationPeriod,
			},
		}),
		invalidatesTags: ["GetBreeds"],
	}),
	deleteBreed: build.mutation<void, number>({
		query: (id) => ({
			url: `/breeds/${id}`,
			method: "DELETE",
		}),
		invalidatesTags: ["GetBreeds"],
	}),
	updateBreed: build.mutation<Breed, { data: UpdateBreedSchema; id: number }>({
		query: ({ data, id }) => ({
			url: `/breeds/${id}`,
			method: "PATCH",
			body: {
				specie: data.specie,
				name: data.name,
				description: data.description,
				typicalMaleWeightRange: data.maleWeightRange,
				typicalFemaleWeightRange: data.femaleWeightRange,
				typicalGestationPeriod: data.gestationPeriod,
			},
		}),
		invalidatesTags: (result, error, id) => [{ type: "GetBreeds", id }],
	}),
});
