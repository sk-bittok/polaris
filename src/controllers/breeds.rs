use axum::{
    Extension, Json, Router,
    body::Body,
    debug_handler,
    extract::{Path, Query, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{delete, get, patch, post},
};
use serde_json::json;

use crate::{
    AppContext, Error, Result,
    middlewares::TokenClaims,
    models::{
        breeds::{Breed, BreedQuery},
        dto::{RegisterBreed, UpdateBreed},
        species::Specie,
        users::User,
    },
    views::animals::BreedResponse,
};

#[debug_handler]
async fn all(
    State(ctx): State<AppContext>,
    Extension(_auth): Extension<TokenClaims>,
    user: User,
    Query(conditions): Query<BreedQuery>,
) -> Result<Response> {
    let breeds = Breed::find_by_all(&ctx.db, user.organisation_pid, &conditions).await?;

    let breeds_futures = breeds
        .iter()
        .map(|breed| {
            let db = ctx.db.clone();
            let breed = breed.clone();
            async move {
                let specie = Specie::find_by_id(&db, breed.specie_id).await?;
                Ok::<_, Error>(BreedResponse::new(&breed, &specie.name))
            }
        })
        .collect::<Vec<_>>();

    let breeds_response = futures::future::try_join_all(breeds_futures).await?;

    Ok((StatusCode::OK, Json(breeds_response)).into_response())
}

#[debug_handler]
async fn one(State(ctx): State<AppContext>, user: User, Path(id): Path<i32>) -> Result<Response> {
    let breed = Breed::find_by_id(&ctx.db, id, user.organisation_pid).await?;
    let specie = Specie::find_by_id(&ctx.db, breed.specie_id).await?;

    Ok((
        StatusCode::OK,
        Json(BreedResponse::new(&breed, &specie.name)),
    )
        .into_response())
}

#[debug_handler]
async fn add(
    State(ctx): State<AppContext>,
    user: User,
    Json(params): Json<RegisterBreed<'static>>,
) -> Result<Response> {
    let mut tx = ctx.db.begin().await?;

    let specie = Specie::find_by_name(&mut *tx, &params.specie).await?;
    let breed = Breed::create(&mut *tx, user.organisation_pid, &params, specie.id).await?;

    tx.commit().await?;

    let response = Response::builder()
        .status(StatusCode::CREATED)
        .body(Body::new(
            json!(BreedResponse::new(&breed, &specie.name)).to_string(),
        ))?;

    Ok(response)
}

#[debug_handler]
async fn remove(
    State(ctx): State<AppContext>,
    user: User,
    Path(id): Path<i32>,
) -> Result<Response> {
    let mut tx = ctx.db.begin().await?;

    let _result = Breed::delete_breed(&mut *tx, user.organisation_pid, id).await?;

    tx.commit().await?;

    Ok((StatusCode::NO_CONTENT, Json(json!({}))).into_response())
}

#[debug_handler]
async fn update(
    State(ctx): State<AppContext>,
    user: User,
    Path(id): Path<i32>,
    Json(params): Json<UpdateBreed<'static>>,
) -> Result<Response> {
    let breed = Breed::update_by_id(&ctx.db, user.organisation_pid, id, &params).await?;

    Ok((StatusCode::CREATED, Json(breed)).into_response())
}

pub fn router(ctx: AppContext) -> Router {
    Router::new()
        .route("/", get(all))
        .route("/", post(add))
        .route("/{id}", get(one))
        .route("/{id}", delete(remove))
        .route("/{id}", patch(update))
        .with_state(ctx)
}
