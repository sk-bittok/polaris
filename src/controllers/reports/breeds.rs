use axum::{
    Json, Router, debug_handler,
    extract::{Query, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{get, post},
};
use serde::{Deserialize, Serialize};

use crate::{
    AppContext, Result,
    models::{BreedSummary, BreedSummaryQuery, users::User},
};

#[derive(Debug, Deserialize, Clone, Serialize)]
struct GenerateReport {
    breed: String,
    specie: String,
}

#[debug_handler]
async fn add(
    State(ctx): State<AppContext>,
    user: User,
    Query(params): Query<GenerateReport>,
) -> Result<Response> {
    let generated = BreedSummary::generate(
        &ctx.db,
        user.organisation_pid,
        &params.breed,
        &params.specie,
    )
    .await?;

    Ok((StatusCode::CREATED, Json(generated)).into_response())
}

#[debug_handler]
async fn all(
    State(ctx): State<AppContext>,
    user: User,
    Query(params): Query<BreedSummaryQuery>,
) -> Result<Response> {
    let reports = BreedSummary::find_list(&ctx.db, user.organisation_pid, &params).await?;

    Ok((StatusCode::OK, Json(reports)).into_response())
}

pub fn router(ctx: AppContext) -> Router {
    Router::new()
        .route("/", post(add))
        .route("/", get(all))
        .with_state(ctx)
}
