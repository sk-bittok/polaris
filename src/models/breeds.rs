#![allow(clippy::missing_errors_doc)]

use std::borrow::Cow;

use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, prelude::FromRow};
use uuid::Uuid;

use super::{ModelError, ModelResult};

#[derive(Debug, Clone, Deserialize)]
pub struct BreedQuery {
    specie: Option<String>,
    name: Option<String>,
}

#[derive(Debug, Deserialize, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegisterBreed<'a> {
    pub(crate) specie: Cow<'a, str>,
    pub(crate) name: Cow<'a, str>,
    pub(crate) description: Option<Cow<'a, str>>,
    pub(crate) typical_male_weight_range: Option<Cow<'a, str>>,
    pub(crate) typical_female_weight_range: Option<Cow<'a, str>>,
    pub(crate) typical_gestation_period: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone, FromRow, Encode)]
pub struct Breed {
    pub(crate) id: i32,
    pub(crate) specie: String,
    pub(crate) name: String,
    pub(crate) description: Option<String>,
    pub(crate) typical_male_weight_range: Option<String>,
    pub(crate) typical_female_weight_range: Option<String>,
    pub(crate) typical_gestation_period: Option<String>,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) organisation_pid: Option<Uuid>,
    pub(crate) is_system_defined: bool,
}

impl Breed {
    pub async fn find_by_id<'e, C>(db: C, id: i32, org_pid: Uuid) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE id = $1 AND (is_system_defined = TRUE OR organisation_pid = $2)")
            .bind(id)
            .bind(org_pid)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_all<'e, C>(
        db: C,
        org_pid: Uuid,
        conditions: &BreedQuery,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let mut query =
            sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE  (is_system_defined = TRUE OR organisation_pid = $1) ORDER BY name")
                .bind(org_pid);

        if let Some(name) = &conditions.name {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE (is_system_defined = TRUE OR organisation_pid = $1) AND name LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{name}%"));
        }

        if let Some(specie) = &conditions.specie {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE organisation_pid = $1 AND specie LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{specie}%",));
        }

        query.fetch_all(db).await.map_err(Into::into)
    }

    pub async fn create<'e, C>(db: C, org_pid: Uuid, dto: &RegisterBreed<'_>) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let query = sqlx::query_as::<_, Self>(
            "INSERT INTO breeds (specie, name, typical_male_weight_range, typical_female_weight_range, typical_gestation_period, organisation_pid, description)
                VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *
            ")
        .bind(&dto.specie)
        .bind(&dto.name)
        .bind(&dto.typical_male_weight_range)
        .bind(&dto.typical_female_weight_range)
        .bind(&dto.typical_gestation_period)
        .bind(org_pid)
        .bind(&dto.description)
        .fetch_one(db)
        .await?;

        Ok(query)
    }
}
