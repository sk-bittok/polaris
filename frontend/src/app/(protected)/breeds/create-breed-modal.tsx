import CustomFormField from "@/components/form-field";
import { Form } from "@/components/ui/form";
import { Category } from "@/models/livestock";
import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import { DialogHeader, DialogTitle, DialogDescription, DialogContent, Dialog, DialogClose, DialogTrigger } from "@/components/ui/dialog";
import { Button } from "@/components/ui/button";
import { registerBreedSchema, RegisterBreedSchema } from "@/lib/schemas/animal";

const speciesOptions = [
  { label: "Cattle", value: Category.Cattle },
  { label: "Sheep", value: Category.Sheep },
  { label: "Goat", value: Category.Goat },
  { label: "Pig", value: Category.Pig },
  { label: "Chicken", value: Category.Chicken },
];

type Props = {
  isOpen: boolean;
  onClose: () => void,
  onCreate: (data: RegisterBreedSchema) => void,
  children: React.ReactNode,
}

export default function CreateBreedModal({ isOpen, onClose, onCreate, children }: Props) {
  const form = useForm<RegisterBreedSchema>({
    resolver: zodResolver(registerBreedSchema),
    mode: 'onChange',
    defaultValues: {
      specie: Category.Cattle,
      name: "",
      description: "",
      maleWeightRange: "",
      femaleWeightRange: "",
      gestationPeriod: "",
    }
  });

  const handleSubmit = (data: RegisterBreedSchema) => {
    onCreate(data);
    form.reset();
    onClose();
  };

  const handleClose = () => {
    form.reset();
    onClose();
  };


  return (
    <Dialog open={isOpen} onOpenChange={(open) => {
      if (!open) {
        handleClose();
      }
    }}>
      <DialogTrigger asChild>{children}</DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>Register new breed</DialogTitle>
          <DialogDescription>Register a new breed. Click save when your are done</DialogDescription>
        </DialogHeader>

        {/* Scrollabel form area */}
        <div className="overflow-y-auto p-2 max-h-[calc(90vh-130px)]">
          <Form {...form}>
            <form onSubmit={form.handleSubmit(handleSubmit)} className="">
              <div className="space-y-4">
                <CustomFormField
                  control={form.control}
                  name="specie"
                  type="select"
                  options={speciesOptions}
                  placeholder="Select a specie"
                  label="Species"
                  inputClassName="w-full"
                />

                <CustomFormField
                  control={form.control}
                  name="name"
                  placeholder="i.e. Hereford, Dorper"
                  label="Name"
                />

                <CustomFormField
                  control={form.control}
                  name="maleWeightRange"
                  placeholder="e.g., 1000-1200 kg"
                  label="Male Weight Range"
                />

                <CustomFormField
                  control={form.control}
                  name="femaleWeightRange"
                  placeholder="e.g., 550-750 kg"
                  label="Female Weight Range"
                />

                <CustomFormField
                  control={form.control}
                  name="gestationPeriod"
                  placeholder="e.g., 278 days"
                  label="Gestation Period"
                />

                <CustomFormField
                  control={form.control}
                  name="description"
                  type="textarea"
                  placeholder="Describe the breed's characteristics, history, purpose, and notable features..."
                  label="Description"
                  inputClassName="min-h-32"
                />
              </div>

              <div className="p-4 border-t border-gray-200 dark:border-gray-700 bg-gray-50 dark:bg-gray-900 mt-4 rounded-lg shadow-md flex items-center justify-end gap-3">
                <Button
                  type="button"
                  onClick={handleClose}
                  className="px-4 py-2 border text-sm font-medium text-white bg-gray-500  dark:text-gray-300 hover:bg-gray-600 dark:hover:bg-gray-700 transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
                >Cancel
                </Button>

                <Button
                  type="submit"
                  className="flex items-center px-4 py-2 text-white text-sm font-medium transition-colors hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-green-600 disabled:opacity-50 disabled:hover:bg-gray-600 disabled:cursor-not-allowed"
                >
                  Create Breed
                </Button>
              </div>
            </form>
          </Form>
        </div>
      </DialogContent>
    </Dialog>
  );
}

