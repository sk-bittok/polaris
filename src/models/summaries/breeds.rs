#![allow(clippy::missing_errors_doc)]
#![allow(clippy::needless_raw_string_hashes)]

use chrono::{DateTime, FixedOffset};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, prelude::FromRow};
use uuid::Uuid;

use crate::models::ModelResult;

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct BreedSummaryQuery {
    pub specie: Option<String>,
    pub breed: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, Encode, FromRow)]
pub struct BreedSummary {
    pub specie_name: String,
    pub breed_name: String,
    pub pid: Uuid,
    pub organisation_pid: Uuid,
    pub id: i32,
    pub specie_id: i32,
    pub breed_id: i32,
    pub total: i32,
    pub males: i32,
    pub females: i32,
    pub unknown_gender: i32,
    pub active: i32,
    pub transferred: i32,
    pub sold: i32,
    pub deceased: i32,
    pub average_age_months: Option<Decimal>,
    pub average_birth_weight_male: Option<Decimal>,
    pub average_birth_weight_female: Option<Decimal>,
    pub average_weight_male: Option<Decimal>,
    pub average_weight_female: Option<Decimal>,
    pub total_purchase_value: Decimal,
    pub average_purchase_price: Decimal,
    pub youngest_age_months: Option<i32>,
    pub oldest_age_months: Option<i32>,
    pub last_calculated_at: DateTime<FixedOffset>,
    pub created_at: DateTime<FixedOffset>,
    pub updated_at: DateTime<FixedOffset>,
}

impl BreedSummary {
    pub async fn generate<'e, C>(
        db: &C,
        org_pid: Uuid,
        breed: &str,
        specie: &str,
    ) -> ModelResult<Self>
    where
        for<'a> &'a C: Executor<'e, Database = Postgres>,
    {
        let data = BreedSummaryExtract::get_from_animals(db, org_pid, breed, specie).await?;

        let query = sqlx::query_as::<_, Self>(
            r#"
            INSERT INTO breed_summary (
                organisation_pid,
                specie_name,
                specie_id,
                breed_name,
                breed_id,
                total,
                males,
                females,
                unknown_gender,
                active,
                transferred,
                sold,
                deceased,
                average_age_months,
                average_weight_male,
                average_weight_female,
                average_birth_weight_male,
                average_birth_weight_female,
                total_purchase_value,
                average_purchase_price,
                youngest_age_months,
                oldest_age_months
            )
            VALUES (
                $1,
                (SELECT name FROM species s WHERE s.name ILIKE $2),
                (SELECT id FROM species s WHERE s.name ILIKE $2),
                (SELECT name FROM breeds b WHERE b.name ILIKE $3),
                (SELECT id FROM breeds b WHERE b.name ILIKE $3),
                $4,
                $5,
                $6,
                $7,
                $8,
                $9,
                $10,
                $11,
                $12,
                $13,
                $14,
                $15,
                $16,
                $17,
                $18,
                $19,
                $20
            )
            RETURNING *
        "#,
        )
        .bind(org_pid)
        .bind(specie)
        .bind(breed)
        .bind(data.as_ref().map_or(0, |data| data.total))
        .bind(data.as_ref().map_or(0, |data| data.males))
        .bind(data.as_ref().map_or(0, |data| data.females))
        .bind(data.as_ref().map_or(0, |data| data.unkown_gender))
        .bind(data.as_ref().map_or(0, |data| data.active))
        .bind(data.as_ref().map_or(0, |data| data.transferred))
        .bind(data.as_ref().map_or(0, |data| data.sold))
        .bind(data.as_ref().map_or(0, |data| data.deceased))
        .bind(data.as_ref().map(|data| data.average_age_months))
        .bind(data.as_ref().map(|data| data.average_weight_male))
        .bind(data.as_ref().map(|data| data.average_weight_female))
        .bind(data.as_ref().map(|data| data.average_birth_weight_male))
        .bind(data.as_ref().map(|data| data.average_birth_weight_female))
        .bind(
            data.as_ref()
                .map_or_else(|| Decimal::new(000, 2), |data| data.total_purchase_value),
        )
        .bind(data.as_ref().map(|data| data.average_purchase_price))
        .bind(data.as_ref().map(|data| data.youngest_age_months))
        .bind(data.as_ref().map(|data| data.oldest_age_months))
        .fetch_one(db)
        .await?;

        Ok(query)
    }

    pub async fn find_list<'e, C>(
        db: &C,
        org_pid: Uuid,
        params: &BreedSummaryQuery,
    ) -> ModelResult<Vec<Self>>
    where
        for<'a> &'a C: Executor<'e, Database = Postgres>,
    {
        let query = sqlx::query_as::<_, Self>(
            r#"
            SELECT * FROM breed_summary bs
            WHERE bs.organisation_pid = $1
                AND ($2 IS NULL OR bs.specie_id = (
                    SELECT id FROM species WHERE name ILIKE $2
                    )
                )
                AND ($3 IS NULL OR bs.breed_id = (
                    SELECT id FROM breeds WHERE name ILIKE $3
                    )
                )
        "#,
        )
        .bind(org_pid)
        .bind(&params.specie)
        .bind(&params.breed)
        .fetch_all(db)
        .await?;

        Ok(query)
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, Encode, FromRow)]
pub struct BreedSummaryExtract {
    pub total: i64,
    pub males: i64,
    pub females: i64,
    pub unkown_gender: i64,
    pub active: i64,
    pub transferred: i64,
    pub sold: i64,
    pub deceased: i64,
    pub average_age_months: Option<Decimal>,
    pub average_weight_male: Option<Decimal>,
    pub average_weight_female: Option<Decimal>,
    pub total_purchase_value: Decimal,
    pub average_birth_weight_male: Option<Decimal>,
    pub average_birth_weight_female: Option<Decimal>,
    pub average_purchase_price: Option<Decimal>,
    pub youngest_age_months: Option<Decimal>,
    pub oldest_age_months: Option<Decimal>,
}

impl BreedSummaryExtract {
    pub async fn get_from_animals<'e, C>(
        db: &C,
        org_pid: Uuid,
        breed: &str,
        specie: &str,
    ) -> ModelResult<Option<Self>>
    where
        for<'a> &'a C: Executor<'e, Database = Postgres>,
    {
        let data = sqlx::query_as::<_, Self>(
            r#"
            SELECT
                COUNT(*)                                                AS      total,
                COUNT(*)    FILTER (WHERE gender = 'male')              AS      males,
                COUNT(*)    FILTER (WHERE gender = 'female')            AS      females,
                COUNT(*)    FILTER (WHERE gender = 'unkown')            AS      unkown_gender,
                COUNT(*)    FILTER (WHERE status = 'active')            AS      active,
                COUNT(*)    FILTER (WHERE status = 'sold')              AS      sold,
                COUNT(*)    FILTER (WHERE status = 'transferred')       AS      transferred,
                COUNT(*)    FILTER (WHERE status = 'deceased')          AS      deceased,
                COALESCE(SUM(purchase_price), 0)                        AS      total_purchase_value,
                AVG(purchase_price)                                     AS      average_purchase_price,
                AVG(current_weight) FILTER (WHERE gender = 'male')      AS      average_weight_male,
                AVG(current_weight) FILTER (WHERE gender = 'female')    AS      average_weight_female,
                AVG(weight_at_birth) FILTER (WHERE gender = 'male')     AS      average_birth_weight_male,
                AVG(weight_at_birth) FILTER (WHERE gender = 'female')   AS      average_birth_weight_female,
                MIN(
                EXTRACT(YEAR FROM AGE(CURRENT_DATE, date_of_birth)) * 12 +
                EXTRACT(MONTH FROM AGE(CURRENT_DATE, date_of_birth))
                )                                                       AS      youngest_age_months,
                MAX(
                EXTRACT(YEAR FROM AGE(CURRENT_DATE, date_of_birth)) * 12 +
                EXTRACT(MONTH FROM AGE(CURRENT_DATE, date_of_birth))
                )                                                       AS      oldest_age_months,
                AVG(
                EXTRACT(YEAR FROM AGE(CURRENT_DATE, date_of_birth)) * 12 +
                EXTRACT(MONTH FROM AGE(CURRENT_DATE, date_of_birth))
                )                                                       AS      average_age_months
            FROM animals
            WHERE
                organisation_pid = $1
            AND
                breed_id = (SELECT id FROM breeds b WHERE b.name ILIKE $2)
            AND
                specie_id = (SELECT id FROM species s WHERE s.name ILIKE $3)
        "#,
        )
            .bind(org_pid)
            .bind(breed)
            .bind(specie)
            .fetch_optional(db)
        .await?;

        Ok(data)
    }
}
