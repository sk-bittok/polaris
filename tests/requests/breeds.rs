use axum::http::StatusCode;
use insta::{Settings, assert_debug_snapshot, with_settings};
use polaris::models::breeds::Breed;
use serial_test::serial;

use crate::request;

use crate::requests::prepare_auth;

macro_rules! configure_insta {
    ($(expr:expr),*) => {
        let mut settings = Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_path("snapshots/breeds");
        settings.set_snapshot_suffix("breeds");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_fetch_all() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db)
            .await
            .inspect_err(|e| eprintln!("Seed error {:?}", e))
            .unwrap();

        let user = prepare_auth::init_login(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server
            .get("/breeds")
            .add_header(auth_header, auth_value)
            .await;

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_uuid().to_vec());
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((request.status_code(), request.text()));
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_fetch_by_id() {
    request(|server, context| async move {
        configure_insta!();

        let user = prepare_auth::init_login(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server
            .get("/breeds/1")
            .add_header(auth_header, auth_value)
            .await;

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_uuid().to_vec());
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((request.status_code(), request.text()));
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_create_one() {
    request(|server, context| async move {
        configure_insta!();

        let post_data = serde_json::json!({
            "name": "Fleckvieh",
            "description": "A dual purpose (milk and meat) breed of cattle with a white head and either a brown, red or yellow with white sections body",
            "typicalGestationPeriod": "285 days",
            "specie": "cattle",
            "typicalMaleWeightRange": "1000-1300",
            "typicalFemaleWeightRange": "600-800"
        });

        let user = prepare_auth::init_login(&server, &context).await;

        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server.post("/breeds")
            .json(&post_data)
            .add_header(auth_header, auth_value).await;

        assert_eq!(request.status_code(), StatusCode::CREATED);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_delete_by_id() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let post_data = serde_json::json!({
            "name": "Fleckvieh",
            "description": "A dual purpose (milk and meat) breed of cattle with a white head and either a brown, red or yellow with white sections body",
            "typicalGestationPeriod": "285 days",
            "specie": "cattle",
            "typicalMaleWeightRange": "1000-1300 kg",
            "typicalFemalWeightRange": "600-800 kg"
        });
        let user = prepare_auth::init_login(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);
        server
            .post("/breeds")
            .json(&post_data)
            .add_header(&auth_header, &auth_value)
            .await;

        let request = server
            .delete("/breeds/8")
            .add_header(auth_header, auth_value)
            .await;

        assert_eq!(request.status_code(), StatusCode::NO_CONTENT);
    }).await;
}

#[tokio::test]
#[serial]
async fn can_update_by_id() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let post_data = serde_json::json!({
            "name": "Fleckvieh",
            "description": "A dual purpose (milk and meat) breed of cattle with a white head and either a brown, red or yellow with white sections body",
            "typicalGestationPeriod": "285 days",
            "specie": "cattle",
            "typicalMaleWeightRange": "1000-1300 kg",
            "typicalFemalWeightRange": "600-800 kg"
        });
        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);
        let _post_request = server
            .post("/breeds")
            .json(&post_data)
            .add_header(&auth_header, &auth_value)
            .await;

        let patch_data = serde_json::json!({
           "description": "Recently, great gains have been made in the genetic milk productiveness of the Fleckvieh through breed management to the point that the 
           Fleckvieh rivals pure milk producing dairy breeds. Special value is placed on the Fleckvieh’s fitness characteristics including fertility, longevity, calving ease, udder health, milking speed, somatic cell count, and persistence. A large percentage of young Fleckvieh cattle are on alpine grazing which results in good overall health and longevity of life. Breeding target: 38 per cent milk, 16 per cent beef, 46 per cent fitness. Fleckvieh cattle exhibit good development and performance capacity along with good conformation. 6,000 kg milk in the 1st lactation and over 7,000 – 9,000 kg milk in later lactations with 4.2 per cent fat and 3.7 per cent protein.
        Beef performance:
            Daily gain: 1.44 kg
            Carcass weight: 57.2 per cent
            EUROP classification E and U: > 85.7 per cent",
            "typicalFemaleWeight": "650-850 kg", 
        });

        let response = server.patch("/breeds/10").json(&patch_data).add_header(&auth_header, &auth_value).await;

        assert_eq!(response.status_code(), StatusCode::CREATED);

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((response.status_code(), response.json::<Breed>()));    
        });

    }).await;
}
