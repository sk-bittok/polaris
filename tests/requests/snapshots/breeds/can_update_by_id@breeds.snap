---
source: tests/requests/breeds.rs
assertion_line: 187
expression: "(response.status_code(), response.json::<Breed>())"
---
(
    201,
    Breed {
        id: ID
        specie_id: ID
        name: "Fleckvieh",
        description: Some(
            "Recently, great gains have been made in the genetic milk productiveness of the Fleckvieh through breed management to the point that the \n           Fleckvieh rivals pure milk producing dairy breeds. Special value is placed on the Fleckvieh’s fitness characteristics including fertility, longevity, calving ease, udder health, milking speed, somatic cell count, and persistence. A large percentage of young Fleckvieh cattle are on alpine grazing which results in good overall health and longevity of life. Breeding target: 38 per cent milk, 16 per cent beef, 46 per cent fitness. Fleckvieh cattle exhibit good development and performance capacity along with good conformation. 6,000 kg milk in the 1st lactation and over 7,000 – 9,000 kg milk in later lactations with 4.2 per cent fat and 3.7 per cent protein.\n        Beef performance:\n            Daily gain: 1.44 kg\n            Carcass weight: 57.2 per cent\n            EUROP classification E and U: > 85.7 per cent",
        ),
        typical_male_weight_range: Some(
            "1000-1300 kg",
        ),
        typical_female_weight_range: None,
        typical_gestation_period: Some(
            "285 days",
        ),
        created_at: DATE,
        updated_at: DATE,
        organisation_pid: Some(
            9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0,
        ),
        is_system_defined: false,
    },
)
